### ISSUE TITLE
Env variable MX_GIT_CACHE_DIR doesn't behave right?

### ISSUE THREAD

**Issue by hlolli** [graalvm/mx#191]
My goal is to compile graal community edition from sources with `mx` and I'm trying to do this with the package manager nix. There's already pre-existing patch for pre-fetching jar files for graal build.  What I want to achieve, is pre-fetching the git repositories like `https://github.com/graalvm/graaljs.git`, gather them into a cache directory and force the mx builder to run the whole thing in offline mode. Without going into more details, it seems to me that the wanted behaviour of the envvar MX_GIT_CACHE_DIR is ignored.

So I have

```
      export MX_GIT_CACHE='refcache'
      export MX_GIT_CACHE_DIR=/nix/store/mx-git-cache
      export DYNAMIC_IMPORTS=/substratevm,/tools,/sulong,/graal-nodejs
      export LIBGRAAL=true
      export GATE="build,test,helloworld"
      mx --primary-suite-path ./vm --java-home=$MY_JAVA_HOME gate --strict-mode --tags $GATE
```

which results in

```
fatal: not a git repository (or any of the parent directories): .git
Fetch from https://github.com/graalvm/graaljs.git into cache /nix/store/mx-git-cache
fatal: not a git repository (or any of the parent directories): .git
git fetch failed
```

bear in mind, that at this point `/nix/store/mx-git-cache/graaljs`, a clone of graaljs.git, exists.

by patching abortOnError to False

```
fatal: not a git repository (or any of the parent directories): .git
Fetch from https://github.com/graalvm/graaljs.git into cache /nix/store/mx-git-cache
fatal: not a git repository (or any of the parent directories): .git
Cloning https://github.com/graalvm/graaljs.git revision 999616fb0efbe0494289b23bbec1e0612fb81201 to /build/graaljs with Git
fatal: repository '/nix/store/mx-git-cache' does not exist
Imported suite 'graal-nodejs' not found (binary or source).
```

This doesn't seem to behave correctly `fatal: repository '/nix/store/mx-git-cache' does not exist` when `/nix/store/mx-git-cache/graaljs` is the directory in which the builder can find `/nix/store/mx-git-cache/graaljs/graal-nodejs`. I can't imagine `~/.mx/git-cache` to be a gihub repository?

In general, patching this build process to work with a purely functional package manager has been quite a challenge. I would suggest allowing a file path to the required dependencies listed in the suite.py files.

https://github.com/oracle/graal/blob/master/vm/mx.vm/suite.py#L10-L16

like here (but it has missing version and that also gives me errors).

**Comment by dougxc** [comment]
@sanzinger can you offer @hlolli any tips on how to his attempts to rely on or use `MX_GIT_CACHE_DIR`?

**Comment by sanzinger** [comment]
Cache mode `refcache` is assuming that the `MX_GIT_CACHE_DIR` is a bare repository. It fetches all remote repositories in this bare repository. So you have multiple remotes in this repository with unrelated roots.

So to solve your problem you just need to point `MX_GIT_CACHE_DIR` to a non existent path (mx initializes with bare repository) or already initialized bare git repository:

First attempt (empty cache):
```
$ mx --primary-suite-path ./vm --java-home=$MY_JAVA_HOME gate --strict-mode --tags $GATE
Initialized empty Git repository in /home/stefan/support/gh_191/graal/mx-git-cache/
Fetch from https://github.com/graalvm/graaljs.git into cache /home/stefan/support/gh_191/graal/mx-git-cache
remote: Enumerating objects: 2303, done.
remote: Counting objects: 100% (2303/2303), done.
remote: Compressing objects: 100% (677/677), done.
remote: Total 408794 (delta 1013), reused 1997 (delta 816), pack-reused 406491
Receiving objects: 100% (408794/408794), 319.64 MiB | 11.00 MiB/s, done.
Resolving deltas: 100% (291341/291341), done.
From https://github.com/graalvm/graaljs
 * [new branch]      master                -> 95c32b191848616e088e24d04ead0ca3da8ec5d3/master
 * [new branch]      release/graal-vm/0.33 -> 95c32b191848616e088e24d04ead0ca3da8ec5d3/release/graal-vm/0.33
 * [new branch]      release/graal-vm/1.0  -> 95c32b191848616e088e24d04ead0ca3da8ec5d3/release/graal-vm/1.0
 * [new tag]         vm-1.0.0-rc15         -> vm-1.0.0-rc15
 * [new tag]         vm-enterprise-0.33    -> vm-enterprise-0.33
 * [new tag]               vm-1.0.0-rc1          -> vm-1.0.0-rc1
 * [new tag]               vm-1.0.0-rc10         -> vm-1.0.0-rc10
 * [new tag]               vm-1.0.0-rc11         -> vm-1.0.0-rc11
 * [new tag]               vm-1.0.0-rc12         -> vm-1.0.0-rc12
 * [new tag]               vm-1.0.0-rc13         -> vm-1.0.0-rc13
 * [new tag]               vm-1.0.0-rc14         -> vm-1.0.0-rc14
 * [new tag]               vm-1.0.0-rc2          -> vm-1.0.0-rc2
 * [new tag]               vm-1.0.0-rc3          -> vm-1.0.0-rc3
 * [new tag]               vm-1.0.0-rc4          -> vm-1.0.0-rc4
 * [new tag]               vm-1.0.0-rc5          -> vm-1.0.0-rc5
 * [new tag]               vm-1.0.0-rc6          -> vm-1.0.0-rc6
 * [new tag]               vm-1.0.0-rc7          -> vm-1.0.0-rc7
 * [new tag]               vm-1.0.0-rc8          -> vm-1.0.0-rc8
 * [new tag]               vm-1.0.0-rc9          -> vm-1.0.0-rc9
 * [new tag]               vm-enterprise-0.31    -> vm-enterprise-0.31
 * [new tag]               vm-enterprise-0.32    -> vm-enterprise-0.32
Cloning https://github.com/graalvm/graaljs.git revision 29c43966086f2d6d0abcf563e37728ac85d4a272 to /home/stefan/support/gh_191/graaljs with Git
Cloning into '/home/stefan/support/gh_191/graaljs'...
done.
Checking out files: 100% (31591/31591), done.
gate: 09 Apr 2019 12:12:59(+00:00) BEGIN: Gate
gate: 09 Apr 2019 12:12:59(+00:00) BEGIN: Versions
Running: mx --primary-suite-path ./vm --java-home= version --oneline
520b94465657  mx 5.215.5
Running: mx --primary-suite-path ./vm --java-home= sversions 
3b15715652d3  vm /home/stefan/support/gh_191/graal
3b15715652d3  truffle /home/stefan/support/gh_191/graal
3b15715652d3  sdk /home/stefan/support/gh_191/graal
29c43966086f  graal-nodejs /home/stefan/support/gh_191/graaljs
29c43966086f  graal-js /home/stefan/support/gh_191/graaljs
3b15715652d3  regex /home/stefan/support/gh_191/graal
3b15715652d3  substratevm /home/stefan/support/gh_191/graal
3b15715652d3  compiler /home/stefan/support/gh_191/graal
3b15715652d3  tools /home/stefan/support/gh_191/graal
3b15715652d3  sulong /home/stefan/support/gh_191/graal
Python version: sys.version_info(major=2, minor=7, micro=16, releaselevel='final', serial=0)
gate: 09 Apr 2019 12:13:00(+00:00) END:   Versions [0:00:00.960389] [disk (free/total): 8.5GB/200.0GB]
gate: 09 Apr 2019 12:13:00(+00:00) BEGIN: JDKReleaseInfo
==== /home/stefan/jdk/labsjdk1.8.0_202-jvmci-0.57 ====
....
```

After removing `graaljs`, it is cloned from the cache without even connecting to the remote:
```
$ mx --primary-suite-path ./vm --java-home=$MY_JAVA_HOME gate --strict-mode --tags $GATE
Cloning https://github.com/graalvm/graaljs.git revision 29c43966086f2d6d0abcf563e37728ac85d4a272 to /home/stefan/support/gh_191/graaljs with Git
Cloning into '/home/stefan/support/gh_191/graaljs'...
done.
Checking out files: 100% (31591/31591), done.
gate: 09 Apr 2019 12:14:54(+00:00) BEGIN: Gate
gate: 09 Apr 2019 12:14:54(+00:00) BEGIN: Versions
Running: mx --primary-suite-path ./vm --java-home= version --oneline
520b94465657  mx 5.215.5
Running: mx --primary-suite-path ./vm --java-home= sversions 
3b15715652d3  vm /home/stefan/support/gh_191/graal
3b15715652d3  truffle /home/stefan/support/gh_191/graal
3b15715652d3  sdk /home/stefan/support/gh_191/graal
29c43966086f  graal-nodejs /home/stefan/support/gh_191/graaljs
29c43966086f  graal-js /home/stefan/support/gh_191/graaljs
3b15715652d3  regex /home/stefan/support/gh_191/graal
3b15715652d3  substratevm /home/stefan/support/gh_191/graal
3b15715652d3  compiler /home/stefan/support/gh_191/graal
3b15715652d3  tools /home/stefan/support/gh_191/graal
3b15715652d3  sulong /home/stefan/support/gh_191/graal
Python version: sys.version_info(major=2, minor=7, micro=16, releaselevel='final', serial=0)
gate: 09 Apr 2019 12:14:55(+00:00) END:   Versions [0:00:00.554744] [disk (free/total): 8.5GB/200.0GB]
gate: 09 Apr 2019 12:14:55(+00:00) BEGIN: JDKReleaseInfo
==== /home/stefan/jdk/labsjdk1.8.0_202-jvmci-0.57 ====
```

**Comment by hlolli** [comment]
Ok I will have to do this differently. The problem I would encounter with the cache initializing a repository, is that it would be writing into a directory belonging to a build dependency, that's somewhat non-posix standard way of building things, but I get the point still of using git. If I copy the depndencies over into a temp build directory I could achieve this, and writing is fine, I try that and let you know how it works out, thanks for a quick reply!

**Comment by hlolli** [comment]
Just while I'm working on this I encountered another totally unrelated bug

```
  File "/build/source/vm/mx.vm/mx_vm.py", line 470, in __init__
    **kw_args)
  File "/build/source/vm/mx.vm/mx_vm.py", line 253, in __init__
    _metadata = self._get_metadata(_sorted_suites)
  File "/build/source/vm/mx.vm/mx_vm.py", line 378, in _get_metadata
    _info = _s.vc.parent_info(_s.vc_dir)
  File "/nix/store/mx-5.215.4/bin/mx.py", line 5925, in parent_info
    out = self.hg_command(vcdir, ["log", "-r", ".", "--template", "{author}|||{date|hgdate}"], abortOnError=abortOnError)
AttributeError: 'HgConfig' object has no attribute 'hg_command'
```

note that I've been patching this file heavily so the line numbers aren't to be taken seriously, but it does really seem to me that hg_command isn't an attribute of the class HgConfig.

**Comment by hlolli** [comment]
Now that I was able to get the cache set up (with patching) so that it can be imported, then later in the build process...
(note I did not trim the logs so please ignore the hashes leading up to directory names)

```
Adding file which is not in the repository: '/nix/store/r3qcv7c9iqw8zr08a9fzjgkpykgv7iwp-mx-git-cache/graaljs/graal-js/mx.graal-js/native-image.properties' in 'native-image.properties'
Process Process-25:
Traceback (most recent call last):
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/multiprocessing/process.py", line 267, in _bootstrap
    self.run()
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/nix/store/r8315qrsyqyw6fwkih25qsl9a9abad8h-mx-5.215.4/bin/mx.py", line 12738, in executeTask
    task.execute()
  File "/nix/store/r8315qrsyqyw6fwkih25qsl9a9abad8h-mx-5.215.4/bin/mx.py", line 964, in execute
    self.build()
  File "/nix/store/r3qcv7c9iqw8zr08a9fzjgkpykgv7iwp-mx-git-cache/graaljs/graal-js/mx.graal-js/mx_graal_js.py", line 114, in build
    unpackIcuData([])
  File "/nix/store/r3qcv7c9iqw8zr08a9fzjgkpykgv7iwp-mx-git-cache/graaljs/graal-js/mx.graal-js/mx_graal_js.py", line 134, in unpackIcuData
    zf.extractall(icu4jDataDir, toExtract)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/zipfile.py", line 1040, in extractall
    self.extract(zipinfo, path, pwd)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/zipfile.py", line 1028, in extract
    return self._extract_member(member, path, pwd)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/zipfile.py", line 1075, in _extract_member
    os.makedirs(upperdirs)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/os.py", line 150, in makedirs
    makedirs(head, mode)
  File "/nix/store/ih03q5vi6kj278l4jk4kh41lqyagzzdi-python-2.7.15/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 13] Permission denied: '/nix/store/r3qcv7c9iqw8zr08a9fzjgkpykgv7iwp-mx-git-cache/graaljs/graal-js/lib'
```

I'm surprised that a directory, that is so carefully version controled with `flock` on top of it, should be considered a writeable directory.

**Comment by sanzinger** [comment]
> Just while I'm working on this I encountered another totally unrelated bug

Seems that the method `hg_command` was removed in 
 d6b4900b5d759704cf13e3ce382a0ba13d7a8874. But do you even use hg repositories in your project? It is legacy code and afaik unlikely to be used.

> Adding file which is not in the repository: '/nix/store/r3qcv7c9iqw8zr08a9fzjgkpykgv7iwp-mx-git-cache/graaljs/graal-js/mx.graal-js/native-image.properties' in 'native-image.properties'

I still don't understand why it is trying to store someting into the git cache? It is not intended to store anything else but just the git repo. I cannot see that this ever happened in our builds.

> I'm surprised that a directory, that is so carefully version controled with flock on top of it, should be considered a writeable directory.

Do you mean the flock usage in the bare git cache repository? It is necessary allow concurrent use.

**Comment by hlolli** [comment]
The hg command is herehttps://github.com/graalvm/mx/blob/master/mx.py#L5925 I got it by initializing fake hg repository to get the builder to read the revision tags. But I fixed it with a patch.

Like I said, I've been patching mx.py to extreame mesures to prevent it from doing any http calls. I'm just refactoring the MX_GIT_CACHE_DIR to behave like I want it to behave :-). Personally I've never encountered a term "git cache repo", it's probably a fine and original solution. As I close this, isn't the `flock` a built in python flock or is it this https://github.com/discoteq/flock ?
